--- Sapi-old.php	2019-04-15 01:44:44.451000000 +0800
+++ Sapi.php	2019-04-15 01:46:42.602000000 +0800
@@ -75,9 +75,15 @@
         if ($contentLength !== null) {
             $output = fopen('php://output', 'wb');
             if (is_resource($body) && get_resource_type($body) == 'stream') {
-                if (PHP_INT_SIZE !== 4){
+                if (PHP_INT_SIZE > 4) {
                     // use the dedicated function on 64 Bit systems
-                    stream_copy_to_stream($body, $output, $contentLength);
+                    // allow PHP to use mmap by copying in 4MiB chunks
+                    $chunk_size = 4 * 1024 * 1024;
+                    stream_set_chunk_size($output, $chunk_size);
+                    $left = $contentLength;
+                    while ($left > 0) {
+                        $left -= stream_copy_to_stream($body, $output, min($left, $chunk_size));
+                    }
                 } else {
                     // workaround for 32 Bit systems to avoid stream_copy_to_stream
                     while (!feof($body)) {
